aten/src/ATen/native/DilatedConvolution.cpp

```diff
@@ -0,0 +1,879 @@
+      T.dim() == DIM && T.size(DIM_SIZE) == SIZE,    \
+      "Need " #T " of dimension ",                   \
+      DIM,                                           \
+      " and " #T ".size[",                           \
+      DIM_SIZE,                                      \
+      "] == ",                                       \
+      SIZE,                                          \
+      " but got input to be of shape ",              \
+      T.sizes())
+
+/*
+  Some convenience macros
+ */
+
+#define INSERT_BATCH_DIMENSION(A, SIZE)        \
+  if (A.defined()) {                           \
+    auto new_sizes = A.sizes().vec();          \
+    new_sizes.insert(new_sizes.begin(), SIZE); \
+    A.resize_(new_sizes);                      \
+  }
+
+#define DROP_BATCH_DIMENSION(A)    \
+  if (A.defined()) {               \
+    A.resize_(A.sizes().slice(1)); \
+  }
+
+namespace at {
+namespace native {
+
+namespace {
+
+inline bool all_positive(IntArrayRef& arr) {
+  return std::all_of(
+      arr.begin(), arr.end(), [](int64_t item) { return item > 0; });
+}
+
+#define OUTPUTSIZE(INDEX, DIM)                                       \
+  (input.size((INDEX) + input.dim() - (DIM)) + 2 * pad_size[INDEX] - \
+   (dilation_size[INDEX] * (kernel_size[INDEX] - 1) + 1)) /          \
+          stride_size[INDEX] +                                       \
+      1
+
+void conv_dilated_shape_check(
+    int64_t dim,
+    const Tensor& input,
+    const Tensor& weight,
+    const Tensor& bias,
+    const Tensor& grad_output,
+    const Tensor& grad_input,
+    const Tensor& grad_weight,
+    const Tensor& grad_bias,
+    IntArrayRef kernel_size,
+    IntArrayRef stride_size,
+    IntArrayRef pad_size,
+    IntArrayRef dilation_size,
+    const Tensor& columns) {
+  TORCH_CHECK(
+      kernel_size.size() == dim,
+      "kernel sizes length should be ",
+      dim,
+      ", but got ",
+      kernel_size.size());
+  TORCH_CHECK(
+      stride_size.size() == dim,
+      "strides length should be ",
+      dim,
+      ", but got ",
+      stride_size.size());
+  TORCH_CHECK(
+      dilation_size.size() == dim,
+      "dilations length should be ",
+      dim,
+      ", but got ",
+      dilation_size.size());
+  TORCH_CHECK(
+      pad_size.size() == dim,
+      "pads length should be ",
+      dim,
+      ", but got ",
+      pad_size.size());
+
+  TORCH_CHECK(
+      all_positive(kernel_size),
+      "kernel size should be greater than zero, but got ",
+      kernel_size);
+  TORCH_CHECK(
+      all_positive(stride_size),
+      "stride should be greater than zero, but got ",
+      stride_size);
+  TORCH_CHECK(
+      all_positive(dilation_size),
+      "dilation should be greater than zero, but got ",
+      dilation_size);
+
+  if (bias.defined()) {
+    TORCH_CHECK_DIM_SIZE(bias, 1, 0, weight.size(0));
+    TORCH_CHECK(bias.is_contiguous(), "bias ought to be contiguous");
+  }
+
+  if (grad_weight.defined()) {
```

Comment:
What happened to the weight shape checks? Previously, shapeCheck only had a single weight parameter, and it was sometimes called with `weight` and sometimes called with `grad_weight` (and that's fine, because these have the same shapes, so the same check works in both situations). In this new code, I only see checks on `grad_weight` and no checks on `weight`.

Reply:
There are the following checks for weight in the new code:
```
TORCH_CHECK(weight.defined(), "weight must be defined");
TORCH_CHECK(weight.dim() == 4, ...);
TORCH_CHECK_DIM_SIZE(bias, 1, 0, weight.size(0)); // when bias is defined
TORCH_CHECK_DIM_SIZE(input, ndim, dimf, weight.size(1));
```
that covers all checks in the old code. However, there exists no checks for `weight.size(i), i>1` that should be equal to `kernel_size[i-2]`.

`grad_weight`, when defined, is guaranteed to have the same shape as `weight`. So, the `grad_weight` shape checks can be deleted.

Reply:
Checking of argument shapes and data structure has been revised now.